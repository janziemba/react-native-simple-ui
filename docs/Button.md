# Button

## Props
Extend React Native's [TouchableOpacity](https://facebook.github.io/react-native/docs/touchableopacity#props) props
- `color` (_string_, default primary) - one of theme colors, [base theme colors](../src/themes/base/colors.js) by default
- `icon` (_object_, default null)
    - `color` (_string_, default null)
    - `colorName` (_string_, default null)
    - `iconSet` (_string_, default `theme.typography.iconSet` which is MaterialIcons by default)
    - `name` (_string_, **required**)
    - `position` (_string_, default left)
    - `size` (_number_, default 24)
- `linearGradient` (_object_, default null)
    - `children` (_node_, default null)
    - `direction` (_string_, default horizontal)
    - `fromColor` (_string_, **required**)
    - `height` (_number_, **required**)
    - `toColor` (_string_, **required**)
    - `width` (_number_, **required**)
- `rounding` (_string_, default normal)
- `size` (_string_, default medium)
- `state` (_string_, default inactive)
- `text` (_string_, **required**)
- `variant` (_string_, default default)

## Examples

### Default

```jsx
<Button
    text="Default"
/>
```

### With different color

```jsx
<Button
    color="success"
    text="With different color"
/>
```

### With icon

```jsx
<Button
    icon={{
        name: 'star',
    }}
    text="With icon"
/>
```

### With icon on right

```jsx
<Button
    icon={{
        name: 'star',
        position: 'right',
    }}
    text="With icon on right"
/>
```

### Rounded

```jsx
<Button
    rounding="rounded"
    text="Rounded"
/>
```

### Sharp

```jsx
<Button
    rounding="sharp"
    text="Sharp"
/>
```

### Small

```jsx
<Button
    size="small"
    text="Small"
/>
```

### Large

```jsx
<Button
    size="large"
    text="Large"
/>
```

### Active

```jsx
<Button
    state="active"
    text="Active"
/>
```

### Disabled

```jsx
<Button
    state="disabled"
    text="Disabled"
/>
```

### Flat

```jsx
<Button
    text="Flat"
    variant="flat"
/>
```

### Outlined

```jsx
<Button
    text="Outlined"
    variant="outlined"
/>
```
