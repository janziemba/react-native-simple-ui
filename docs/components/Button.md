---
id: button
title: Button
---
## Props
Extend React Native's [TouchableOpacity](https://facebook.github.io/react-native/docs/touchableopacity#props) props
- `color` (_string_, default primary) - one of theme colors, [base theme colors](../Theme.md#colors) by default
- `disabled` (_boolean_, default false)
- `icon` (_object_, default null)
    - [`Icon` props](Icon.md#props)
    - `position` (_string_, default left, only matters if a `text` is set)
- `linearGradient` (_object_, default null)
    - [`LinearGradient` props](LinearGradient.md#props) except `height` and `width` which are calculated automatically based on a container
- `loading` (_boolean_, default false)
- `rounding` (_string_, default normal)
- `size` (_string_, default medium) - one of theme spacing, [base theme spacing](../Theme.md#spacing) by default
- `text` (_string_, default null)
- `variant` (_string_, default default)

<img src="../assets/Button/1.png" alt="Button" width="250" style="display: inline-block;">
<img src="../assets/Button/2.png" alt="Button" width="250" style="display: inline-block;">
<img src="../assets/Button/3.png" alt="Button" width="250" style="display: inline-block;">
<img src="../assets/Button/4.png" alt="Button" width="250" style="display: inline-block;">
<img src="../assets/Button/5.png" alt="Button" width="250" style="display: inline-block;">

## How do styles work
The styles are merged in this order: base, rounding, size, color, variant and iconPosition. It means basic styles
are applied, then rounded borders, then then the button is resized, then the button gets its colors - the color is
applied for both background and text. Then variant decides if the color should be applied for background only
(default variant), for text only (flat variant) or for text and border (outlined variant). Then iconPosition decides
which flexDirection and padding should be applied to container and iconContainer.

## Examples

### Default

```jsx
<Button
    text="With text"
/>
```

```jsx
<Button
    icon={{
        name: 'star',
    }}
/>
```

### Color

```jsx
<Button
    color="success"
    text="With different color"
/>
```

### Icon

```jsx
<Button
    icon={{
        name: 'star',
    }}
    text="With icon"
/>
```

```jsx
<Button
    icon={{
        name: 'star',
        position: 'right',
    }}
    text="With icon on right"
/>
```

### Rounding

```jsx
<Button
    rounding="rounded"
    text="Rounded"
/>
```

```jsx
<Button
    rounding="sharp"
    text="Sharp"
/>
```

### Size

```jsx
<Button
    size="small"
    text="Small"
/>
```

```jsx
<Button
    size="large"
    text="Large"
/>
```

### Variant

```jsx
<Button
    text="Flat"
    variant="flat"
/>
```

```jsx
<Button
    text="Outlined"
    variant="outlined"
/>
```

### States

```jsx
<Button
    disabled
    text="Disabled"
/>
```

```jsx
<Button
    loading
/>
```
