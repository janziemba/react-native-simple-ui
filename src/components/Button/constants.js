// @flow

import colors from '../../themes/base/colors';
import type { ObjectOfStringsType } from '../../types';

export const COLORS: ObjectOfStringsType = {
    ...colors.palette,
    white: 'white',
};

export const ICON_POSITIONS: ObjectOfStringsType = {
    bottom: 'bottom',
    left: 'left',
    right: 'right',
    top: 'top',
};

export const ROUNDING: ObjectOfStringsType = {
    normal: 'normal',
    rounded: 'rounded',
    sharp: 'sharp',
};

export const SIZES: ObjectOfStringsType = {
    large: 'large',
    medium: 'medium',
    small: 'small',
};

export const STATES: ObjectOfStringsType = {
    active: 'active',
    disabled: 'disabled',
    inactive: 'inactive',
};

export const VARIANTS: ObjectOfStringsType = {
    default: 'default',
    flat: 'flat',
    outlined: 'outlined',
};
