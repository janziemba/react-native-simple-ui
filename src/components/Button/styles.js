// @flow

import { StyleSheet } from 'react-native';

import merge from 'lodash/merge';

import type { ObjectOfStringsType, ObjectOfObjectsType, ThemeShapeType } from '../../types';
import darkenOrLighten from '../../utils/darkenOrLighten';
import { COLORS, ICON_POSITIONS, ROUNDING, SIZES, STATES, VARIANTS } from './constants';

export type StylesType = {
    base: {
        container: StyleSheet.Styles,
        text: StyleSheet.Styles,
        touchable: StyleSheet.Styles,
    },
    colors: ObjectOfObjectsType,
    iconPositions: ObjectOfObjectsType,
    rounding: ObjectOfObjectsType,
    sizes: ObjectOfObjectsType,
    variants: ObjectOfObjectsType,
};

export default ({ colors, radiuses, spacing, typography }: ThemeShapeType): StylesType => {
    const allColors: ObjectOfStringsType = merge({}, COLORS, colors.palette);
    const colorOptions: ObjectOfObjectsType = {};

    Object.keys(allColors).forEach((colorName: string) => {
        const color = colors.palette[colorName] || colorName;

        colorOptions[colorName] = {
            [STATES.active]: {
                container: {
                    backgroundColor: darkenOrLighten(color),
                    borderColor: darkenOrLighten(color),
                },
                text: {
                    color: darkenOrLighten(color),
                },
            },
            [STATES.disabled]: {
                container: {
                    backgroundColor: darkenOrLighten(color, 0.5),
                    borderColor: darkenOrLighten(color, 0.5),
                },
                text: {
                    color: darkenOrLighten(color, 0.5),
                },
            },
            [STATES.inactive]: {
                container: {
                    backgroundColor: color,
                    borderColor: color,
                },
                text: {
                    color,
                },
            },
        };
    });

    return {
        base: {
            container: {
                ...StyleSheet.absoluteFillObject,
                alignItems: 'center',
                borderWidth: 1,
                justifyContent: 'center',
                overflow: 'hidden',
            },
            text: {
                fontWeight: typography.fontWeights.medium,
                textAlign: 'center',
            },
            touchable: {
                alignSelf: 'stretch',
                position: 'relative',
            },
        },
        colors: colorOptions,
        iconPositions: {
            [ICON_POSITIONS.bottom]: {
                iconContainer: {
                    marginTop: spacing.small,
                },
            },
            [ICON_POSITIONS.left]: {
                container: {
                    flexDirection: 'row',
                },
                iconContainer: {
                    marginRight: spacing.small,
                },
            },
            [ICON_POSITIONS.right]: {
                container: {
                    flexDirection: 'row',
                },
                iconContainer: {
                    marginLeft: spacing.small,
                },
            },
            [ICON_POSITIONS.top]: {
                iconContainer: {
                    marginBottom: spacing.small,
                },
            },
        },
        rounding: {
            [ROUNDING.normal]: {
                borderRadius: radiuses.small,
            },
            [ROUNDING.rounded]: {
                borderRadius: 1000,
            },
            [ROUNDING.sharp]: {
                borderRadius: 0,
            },
        },
        sizes: {
            [SIZES.large]: {
                container: {
                    paddingHorizontal: spacing.huge,
                    paddingVertical: spacing.large,
                },
                text: {
                    fontSize: typography.fontSizes.large,
                },
            },
            [SIZES.medium]: {
                container: {
                    paddingHorizontal: spacing.large,
                    paddingVertical: spacing.medium,
                },
                text: {
                    fontSize: typography.fontSizes.medium,
                },
            },
            [SIZES.small]: {
                container: {
                    paddingHorizontal: spacing.medium,
                    paddingVertical: spacing.small,
                },
                text: {
                    fontSize: typography.fontSizes.small,
                },
            },
        },
        variants: {
            [VARIANTS.default]: {
                text: {
                    color: 'white',
                },
            },
            [VARIANTS.flat]: {
                container: {
                    backgroundColor: 'transparent',
                    borderColor: 'transparent',
                },
            },
            [VARIANTS.outlined]: {
                container: {
                    backgroundColor: 'transparent',
                },
            },
        },
    };
};
