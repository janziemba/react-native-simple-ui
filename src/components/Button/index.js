// @flow

import React, { PureComponent } from 'react';
import type { Node } from 'react';
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native';

import omit from 'lodash/omit';
import type { Props as TouchableWithoutFeedbackProps } from 'react-native/Libraries/Components/Touchable/TouchableWithoutFeedback';

import withTheme from '../../themes/withTheme';
import hasStyleChanged from '../../utils/hasStyleChanged';
import Icon from '../Icon';
import type { Props as IconProps } from '../Icon';
import type { Props as LinearGradientProps } from '../LinearGradient';
import { ICON_POSITIONS, ROUNDING, SIZES, STATES, VARIANTS } from './constants';
import styles from './styles';
import type { StylesType } from './styles';

type Props = TouchableWithoutFeedbackProps & {
    color: string,
    icon: IconProps,
    linearGradient: LinearGradientProps,
    rounding: string, // eslint-disable-line react/no-unused-prop-types, max-len
    size: string, // eslint-disable-line react/no-unused-prop-types
    styles: StylesType, // eslint-disable-line react/no-unused-prop-types, max-len
    text: string,
    variant: string, // eslint-disable-line react/no-unused-prop-types, max-len
};

const defaultProps = {
    ...TouchableOpacity.defaultProps,
    color: 'primary',
    icon: null,
    linearGradient: null,
    rounding: ROUNDING.normal,
    size: SIZES.medium,
    variant: VARIANTS.default,
};

type MergedStylesType = {
    container: Array<StyleSheet.Styles>,
    iconColor: string,
    iconContainer: StyleSheet.Styles,
    text: StyleSheet.Styles,
    touchable: StyleSheet.Styles,
};

type State = {
    active: boolean,
    styles: MergedStylesType,
};

const getStyles = (props: Props, state?: State): MergedStylesType => {
    const { color, disabled, icon, rounding, size, styles, variant } = props;

    let colorState: string = STATES.inactive;
    if (disabled) {
        colorState = STATES.disabled;
    } else if (state && state.active) {
        colorState = STATES.active;
    }

    return {
        container: [
            styles.base.container,
            styles.rounding[rounding],
            styles.sizes[size].container,
            styles.colors[color][colorState].container,
            styles.variants[variant].container,
            icon ? styles.iconPositions[icon.position || ICON_POSITIONS.left].container : {},
        ],
        iconColor: (styles.variants[variant].text && styles.variants[variant].text.color)
            || styles.colors[color][colorState].text.color
            || styles.base.text.color,
        iconContainer: [
            styles.base.iconContainer,
            icon ? styles.iconPositions[icon.position || ICON_POSITIONS.left].iconContainer : {},
        ],
        text: [
            styles.base.text,
            styles.sizes[size].text,
            styles.colors[color][colorState].text,
            styles.variants[variant].text,
        ],
        touchable: styles.touchable,
    };
};

class Button extends PureComponent<Props, State> {
    static defaultProps = defaultProps;
    state = {
        active: false, // eslint-disable-line react/no-unused-state
        styles: getStyles(this.props),
    };
    constructor(props: Props): void {
        super(props);

        (this: any).onPressIn = this.onPressIn.bind(this);
        (this: any).onPressOut = this.onPressOut.bind(this);
        (this: any).renderIcon = this.renderIcon.bind(this);
        (this: any).renderIconAfterText = this.renderIconAfterText.bind(this);
        (this: any).renderIconBeforeText = this.renderIconBeforeText.bind(this);
    }
    componentWillReceiveProps(nextProps: Props): void {
        const propsOnWhichDependsTheStyle: Array<string> = ['color', 'disabled', 'icon', 'rounding', 'size', 'variant'];

        if (hasStyleChanged(propsOnWhichDependsTheStyle, nextProps, this.props)) {
            this.setState(prevState => ({ styles: getStyles(nextProps, prevState) }));
        }
    }
    onPressIn(): void {
        this.setState(
            // eslint-disable-next-line react/no-unused-state
            { active: true },
            () => this.setState(prevState => ({ styles: getStyles(this.props, prevState) })),
        );
    }
    onPressOut(): void {
        this.setState(
            // eslint-disable-next-line react/no-unused-state
            { active: false },
            () => this.setState(prevState => ({ styles: getStyles(this.props, prevState) })),
        );
    }
    renderIcon(): Node {
        const { icon } = this.props;
        const { styles } = this.state;

        if (!icon) {
            return null;
        }

        return (
            <View style={styles.iconContainer}>
                <Icon
                    color={styles.iconColor}
                    iconSet={icon.iconSet}
                    name={icon.name}
                    size={icon.size || 16}
                />
            </View>
        );
    }
    renderIconAfterText(): Node {
        const { icon } = this.props;

        if (!icon || !icon.position
            || ![ICON_POSITIONS.bottom, ICON_POSITIONS.right].includes(icon.position)) {
            return null;
        }

        return this.renderIcon();
    }
    renderIconBeforeText(): Node {
        const { icon } = this.props;

        if (!icon
            || ![ICON_POSITIONS.left, ICON_POSITIONS.top, undefined].includes(icon.position)) {
            return null;
        }

        return this.renderIcon();
    }
    render(): Node {
        const { disabled, onPress, text } = this.props;
        const { styles } = this.state;

        return (
            <TouchableOpacity
                {...omit(this.props, ['styles'])}
                accessibilityRole="button"
                activeOpacity={1}
                delayPressOut={50} // to highlight the button even if it's touched quickly
                disabled={disabled || !onPress}
                onPressIn={this.onPressIn}
                onPressOut={this.onPressOut}
                style={styles.touchable}
            >
                <View style={styles.container}>
                    {this.renderIconBeforeText()}
                    <Text style={styles.text}>
                        {text}
                    </Text>
                    {this.renderIconAfterText()}
                </View>
            </TouchableOpacity>
        );
    }
}

Button = withTheme(styles, 'Button')(Button);

export default Button;
