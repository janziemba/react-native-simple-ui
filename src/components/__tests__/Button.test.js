import React from 'react';
import renderer from 'react-test-renderer';

import Button from '../Button';

it('renders correctly', () => {
    const tree = renderer.create(
        <Button text="Lorem ipsum" />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
