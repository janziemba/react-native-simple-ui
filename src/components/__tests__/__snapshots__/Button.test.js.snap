// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessibilityRole="button"
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderWidth": 1,
          "justifyContent": "center",
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 4,
        },
        Object {
          "paddingHorizontal": 16,
          "paddingVertical": 12,
        },
        Object {
          "backgroundColor": "#158cba",
          "borderColor": "#158cba",
        },
        undefined,
        Object {},
      ]
    }
  >
    <Text>
      Lorem ipsum
    </Text>
  </View>
</View>
`;
